crates/next-core/src/manifest.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::HashSet;
+
+use anyhow::Result;
+use mime::APPLICATION_JSON;
+use turbo_tasks::primitives::StringsVc;
+use turbo_tasks_fs::File;
+use turbopack_core::asset::AssetContentVc;
+use turbopack_dev_server::source::{
+    combined::CombinedContentSourceVc, conditional::ConditionalContentSourceVc, ContentSource,
+    ContentSourceContent, ContentSourceData, ContentSourceResultVc, ContentSourceVc,
+};
+
+use crate::nodejs::{
+    node_api_source::NodeApiContentSourceVc, node_rendered_source::NodeRenderContentSourceVc,
+};
+
+#[turbo_tasks::value(shared)]
+pub struct DevManifestContentSource {
+    pub page_roots: Vec<ContentSourceVc>,
+}
+
+#[turbo_tasks::value_impl]
+impl DevManifestContentSourceVc {
+    #[turbo_tasks::function]
+    async fn find_routes(self) -> Result<StringsVc> {
+        let this = &*self.await?;
+        let mut queue = this.page_roots.clone();
+        let mut routes = HashSet::new();
```

Comment:
since this is iterated below and has unpredictable order, `IndexSet`?

Reply:
ah, good point

Reply:
done